File: deploy/release/clean-v1-release.py
#!/usr/bin/env python3
"""Read the non-secret, canonical clean-v1 release record on an operator host."""

import json
import os
import re
import stat
import sys
from datetime import datetime, timezone

MAX_BYTES = 16 * 1024
SHA256 = re.compile(r"^[0-9a-f]{64}$")
SOURCE_SHA = re.compile(r"^[0-9a-f]{40}$")
IMAGE = re.compile(r"^[a-z0-9][a-z0-9.-]*(?:/[a-z0-9][a-z0-9._-]*)+@sha256:[0-9a-f]{64}$")
VERSION = re.compile(r"^[0-9]+\.[0-9]+\.[0-9]+(?:-[0-9A-Za-z.-]+)?(?:\+[0-9A-Za-z.-]+)?$")
RELEASE_ID = re.compile(r"^clean-v1-[a-z0-9][a-z0-9-]{2,63}$")
TIMESTAMP = re.compile(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$")
URL = re.compile(r"^https://[^\s?#]+(?:[?#][^\s]*)?$")


def fail(message):
    raise ValueError("clean-v1 release record: " + message)


def exact_keys(value, keys, path):
    if not isinstance(value, dict) or sorted(value) != sorted(keys):
        fail(path + " must contain exactly: " + ", ".join(sorted(keys)))


def string(value, name, pattern):
    if not isinstance(value, str) or not pattern.fullmatch(value):
        fail(name + " is invalid")
    return value


def timestamp(value, name):
    string(value, name, TIMESTAMP)
    try:
        parsed = datetime.strptime(value, "%Y-%m-%dT%H:%M:%SZ").replace(tzinfo=timezone.utc)
    except ValueError:
        fail(name + " must be a UTC second timestamp")
    if parsed.strftime("%Y-%m-%dT%H:%M:%SZ") != value:
        fail(name + " must be a UTC second timestamp")
    return value


def validate(value):
    exact_keys(value, ["authority_image", "baseline_compatibility_class", "kind", "person_client", "release_id", "released_at", "schema_version", "source_sha"], "$")
    if type(value["schema_version"]) is not int or value["schema_version"] != 1:
        fail("schema_version must equal integer 1")
    if value["kind"] != "echo-clean-v1-release":
        fail("kind must be echo-clean-v1-release")
    string(value["release_id"], "release_id", RELEASE_ID)
    timestamp(value["released_at"], "released_at")
    if value["baseline_compatibility_class"] != "clean-v1":
        fail("baseline_compatibility_class must equal clean-v1")
    string(value["source_sha"], "source_sha", SOURCE_SHA)
    exact_keys(value["authority_image"], ["reference"], "authority_image")
    string(value["authority_image"]["reference"], "authority_image.reference", IMAGE)
    exact_keys(value["person_client"], ["artifact_sha256", "artifact_url", "package", "version"], "person_client")
    if value["person_client"]["package"] != "@echo-brain/person-client":
        fail("person_client.package must equal @echo-brain/person-client")
    string(value["person_client"]["version"], "person_client.version", VERSION)
    string(value["person_client"]["artifact_url"], "person_client.artifact_url", URL)
    string(value["person_client"]["artifact_sha256"], "person_client.artifact_sha256", SHA256)
    return value


def read(path):
    state = os.lstat(path)
    if not stat.S_ISREG(state.st_mode) or state.st_size <= 0 or state.st_size > MAX_BYTES:
        fail("record must be a non-empty regular file no larger than 16 KiB")
    with open(path, "r", encoding="utf-8") as source:
        raw = source.read()
    try:
        value = validate(json.loads(raw))
    except json.JSONDecodeError:
        fail("record is not valid JSON")
    canonical = json.dumps(value, ensure_ascii=False, sort_keys=True, separators=(",", ":")) + "\n"
    if raw != canonical:
        fail("record bytes are not canonical JSON followed by one newline")
    return value


def main(argv):
    if len(argv) not in (2, 3) or argv[0] not in ("validate", "field") or (argv[0] == "field" and len(argv) != 3):
        fail("usage: clean-v1-release.py <validate|field> <record> [authority-image|baseline-class|client-url|client-sha256|client-version]")
    record = read(argv[1])
    if argv[0] == "validate":
        sys.stdout.write(json.dumps(record, ensure_ascii=False, sort_keys=True, separators=(",", ":")) + "\n")
        return
    fields = {
        "authority-image": record["authority_image"]["reference"],
        "baseline-class": record["baseline_compatibility_class"],
        "release-id": record["release_id"],
        "client-url": record["person_client"]["artifact_url"],
        "client-sha256": record["person_client"]["artifact_sha256"],
        "client-version": record["person_client"]["version"],
        "source-sha": record["source_sha"],
    }
    if argv[2] not in fields:
        fail("unknown field")
    sys.stdout.write(fields[argv[2]] + "\n")


try:
    main(sys.argv[1:])
except (OSError, ValueError) as error:
    sys.stderr.write(str(error) + "\n")
    sys.exit(1)

File: tools/clean-v1-release.mjs
#!/usr/bin/env node

import { lstatSync, readFileSync, writeFileSync } from 'node:fs';
import { resolve } from 'node:path';
import process from 'node:process';

const MAX_RELEASE_RECORD_BYTES = 16 * 1024;
const SHA256 = /^[0-9a-f]{64}$/;
const SOURCE_SHA = /^[0-9a-f]{40}$/;
const IMAGE = /^[a-z0-9][a-z0-9.-]*(?:\/[a-z0-9][a-z0-9._-]*)+@sha256:[0-9a-f]{64}$/;
const VERSION = /^[0-9]+\.[0-9]+\.[0-9]+(?:-[0-9A-Za-z.-]+)?(?:\+[0-9A-Za-z.-]+)?$/;

function fail(message) {
  throw new Error(`clean-v1 release record: ${message}`);
}

function object(value, path) {
  if (value === null || typeof value !== 'object' || Array.isArray(value)) {
    fail(`${path} must be an object`);
  }
  return value;
}

function exactKeys(value, keys, path) {
  const actual = Object.keys(value).sort();
  const expected = [...keys].sort();
  if (JSON.stringify(actual) !== JSON.stringify(expected)) {
    fail(`${path} must contain exactly: ${expected.join(', ')}`);
  }
}

function text(value, path, pattern) {
  if (typeof value !== 'string' || !pattern.test(value)) {
    fail(`${path} is invalid`);
  }
  return value;
}

function timestamp(value, path) {
  let parsed;
  if (
    typeof value !== 'string' ||
    !/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$/.test(value)
  ) {
    fail(`${path} must be a UTC second timestamp`);
  }
  parsed = new Date(value);
  if (
    !Number.isFinite(parsed.getTime()) ||
    parsed.toISOString() !== `${value.slice(0, -1)}.000Z`
  ) {
    fail(`${path} must be a UTC second timestamp`);
  }
  return value;
}

/** RFC 8785 key ordering is sufficient for this all-primitive release record. */
export function canonicalJson(value) {
  if (value === null) return 'null';
  if (typeof value === 'string') return JSON.stringify(value);
  if (typeof value === 'number') {
    if (!Number.isFinite(value)) fail('contains a non-finite number');
    return JSON.stringify(value);
  }
  if (typeof value === 'boolean') return value ? 'true' : 'false';
  if (Array.isArray(value)) return `[${value.map(canonicalJson).join(',')}]`;
  const record = object(value, '$');
  return `{${Object.keys(record)
    .sort()
    .map((key) => `${JSON.stringify(key)}:${canonicalJson(record[key])}`)
    .join(',')}}`;
}

export function validateCleanV1Release(value) {
  const record = object(value, '$');
  exactKeys(
    record,
    [
      'authority_image',
      'baseline_compatibility_class',
      'kind',
      'person_client',
      'release_id',
      'released_at',
      'schema_version',
      'source_sha',
    ],
    '$',
  );
  if (!Number.isInteger(record.schema_version) || record.schema_version !== 1) {
    fail('schema_version must equal integer 1');
  }
  if (record.kind !== 'echo-clean-v1-release') fail('kind must be echo-clean-v1-release');
  const releaseId = text(record.release_id, 'release_id', /^clean-v1-[a-z0-9][a-z0-9-]{2,63}$/);
  const releasedAt = timestamp(record.released_at, 'released_at');
  if (record.baseline_compatibility_class !== 'clean-v1') {
    fail('baseline_compatibility_class must equal clean-v1');
  }
  const sourceSha = text(record.source_sha, 'source_sha', SOURCE_SHA);

  const authorityImage = object(record.authority_image, 'authority_image');
  exactKeys(authorityImage, ['reference'], 'authority_image');
  const authorityReference = text(authorityImage.reference, 'authority_image.reference', IMAGE);

  const personClient = object(record.person_client, 'person_client');
  exactKeys(personClient, ['artifact_sha256', 'artifact_url', 'package', 'version'], 'person_client');
  if (personClient.package !== '@echo-brain/person-client') {
    fail('person_client.package must equal @echo-brain/person-client');
  }
  const clientVersion = text(personClient.version, 'person_client.version', VERSION);
  const artifactUrl = text(personClient.artifact_url, 'person_client.artifact_url', /^https:\/\/[^\s?#]+(?:[?#][^\s]*)?$/);
  const artifactSha256 = text(personClient.artifact_sha256, 'person_client.artifact_sha256', SHA256);

  return Object.freeze({
    schema_version: 1,
    kind: 'echo-clean-v1-release',
    release_id: releaseId,
    released_at: releasedAt,
    baseline_compatibility_class: 'clean-v1',
    source_sha: sourceSha,
    authority_image: Object.freeze({ reference: authorityReference }),
    person_client: Object.freeze({
      package: '@echo-brain/person-client',
      version: clientVersion,
      artifact_url: artifactUrl,
      artifact_sha256: artifactSha256,
    }),
  });
}

export function readCleanV1Release(path) {
  const absolute = resolve(path);
  const state = lstatSync(absolute);
  if (state.isSymbolicLink() || !state.isFile() || state.size <= 0 || state.size > MAX_RELEASE_RECORD_BYTES) {
    fail('record must be a non-empty regular file no larger than 16 KiB');
  }
  const raw = readFileSync(absolute, 'utf8');
  let parsed;
  try {
    parsed = JSON.parse(raw);
  } catch {
    fail('record is not valid JSON');
  }
  const validated = validateCleanV1Release(parsed);
  if (raw !== `${canonicalJson(validated)}\n`) {
    fail('record bytes are not canonical JSON followed by one newline');
  }
  return validated;
}

function usage() {
  return 'usage: clean-v1-release.mjs <create|validate|field> <record> [output-or-field]';
}

function main(argv) {
  const [command, path, field] = argv;
  if (command !== 'create' && command !== 'validate' && command !== 'field') fail(usage());
  if (typeof path !== 'string' || (command !== 'validate' && typeof field !== 'string')) fail(usage());
  if (command === 'create') {
    let draft;
    try {
      draft = JSON.parse(readFileSync(resolve(path), 'utf8'));
    } catch {
      fail('draft is not valid JSON');
    }
    const record = validateCleanV1Release(draft);
    try {
      writeFileSync(resolve(field), `${canonicalJson(record)}\n`, { encoding: 'utf8', mode: 0o600, flag: 'wx' });
    } catch (error) {
      fail(`could not create canonical record: ${error instanceof Error ? error.message : String(error)}`);
    }
    process.stdout.write(`${resolve(field)}\n`);
    return;
  }
  const record = readCleanV1Release(path);
  if (command === 'validate') {
    process.stdout.write(`${canonicalJson(record)}\n`);
    return;
  }
  const fields = {
    'authority-image': record.authority_image.reference,
    'baseline-class': record.baseline_compatibility_class,
    'release-id': record.release_id,
    'client-url': record.person_client.artifact_url,
    'client-sha256': record.person_client.artifact_sha256,
    'client-version': record.person_client.version,
    'source-sha': record.source_sha,
  };
  if (!Object.hasOwn(fields, field)) fail(usage());
  process.stdout.write(`${fields[field]}\n`);
}

try {
  main(process.argv.slice(2));
} catch (error) {
  process.stderr.write(`${error instanceof Error ? error.message : String(error)}\n`);
  process.exitCode = 1;
}
